src/main/java/cryptator/specs/ICryptaNode.java

```diff
@@ -39,42 +39,44 @@ public interface ICryptaNode {
      * @return the left child
      */
     ICryptaNode getLeftChild();
-    // FIXME Use java.util.Optional ?
 
     /**
      * Gets the right child if any.
      *
      * @return the right child
      */
     ICryptaNode getRightChild();
-    // FIXME Use java.util.Optional ?
 
     /**
-     * Checks if the node is a constant leaf of the tree.
+     * Checks if the node is an internal node of the tree.
      *
-     * @return true, if it is a constant leaf
+     * @return true, if it is internal node, otherwise it is a leaf
      */
-    boolean isConstantLeaf();
-
+    boolean isInternalNode();
+    
+    
     /**
-     * Checks if the node is a word leaf of the tree.
+     * Checks if the subtree is a constant of the tree.
      *
-     * @return true, if it is a word leaf
+     * @return true, if it is constant, otherwise false.
      */
-    boolean isWordLeaf();
+    boolean isConstant();
```

Comment:
Little question.
I see why you return true if the subtree is made uniquely by constants. 
But, how can I get back the value (the constant) of this particular internal node in an easy way (for constant we can do `Interger.parseInt(new String(n.getWord))`?
And in what situation is it usefull to know that a subtree has no leaves of type Word?
For example further in the code, we always check if the node `isConstant` in branches corresponding to the `!isInternalNode` case (see `CryptaEvaluation` and `CryptaFeatures`). 
I make this remark, because I'm still thinking to an enum returning the type of the node to use a switch where needed intead of nested conditional if branching. 

EDIT:
I have an idea to keep both your impl and what I would like to do: (the CONSTANT_INTERNAL_NODE is here if you think it is good to have this informations on subtrees)
```java
enum NodeType {
CONSTANT_LEAF(true), WORD_LEAF(true), CONSTANT_INTERNAL_NODE(false), CLASSIC_INTERNAL_NODE(false);
public final boolean isInternalNode;
NodeType(boolean isNode) {this.isInternalNode=isNode}
}

ICryptaNode += {
NodeType getNodeTypology();

default boolean isInternalNode(){ return getNodeTypology().isInternalNode }
default boolean isLeaf(){ return !isInteralNode() }
}
```
and with this we will be able to do something like this (code example for CryptaFetaures, we can do the same for CryptaEvaluation)
```java
switch node.getNodeTYpology(){
 case CONSTANT_LEAF: constantCount++; break;
 case WORD_LEAF: accept(node.getWord()): break;
 case CONSTANT_INTERNAL_NODE:
 case CLASSIC_INTERNAL_NODE: operators.add(node.getOperator());
}
```
I don't have checked if the code will compile and work in java, it is just an idea to show you what I'm thinking on. The enum is not meant to do a kind of instanceof, but just to replace the *if*s in the code. Inside the `case`s we will use polymorphism to compute the wanted behavour.

I hope I have clearly explained my thought 🤔

Reply:
The solution with enums is also doable, but here are my con arguments. 
- I do not find the swith shorter or more readable than the simple nested if.
- We have to store some kind of type information as an attribute. 
- Enum makes inheritance more difficult. We cannot simply extend the enum to handle a third property `isXXX`.  

> But, how can I get back the value (the constant) of this particular internal node in an easy way (for constant we can do Interger.parseInt(new String(n.getWord))?

You have to use the evaluator. `isConstant` indicates if you can evaluate the subtree BEFORE solving the cryptarithm.

I suggest that we discuss these alternatives during the next session.